File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.js
// function solution(s){
//   var answer = true;
    
//   while(s.includes('()')){
//     s=s.split('()').join('')
//   }
//   if(s!=''){answer=false}
//   return answer;
//     //시간초과로 실패 
// }
function solution(s){
  var answer = true;
  //(())
  //()()
  //()(())
  let sum=0
  let obj = {'(':1,')':-1}

  for (i of s){
    sum+=Number(obj[i])

    if (sum <= -1){
      break
    }
  }
  if (sum!==0){
    answer=false
  }
  return answer;
}

File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    memo=[]
    for i in s: 
        if not memo and i==')':
            return False
        if i=='(':
            memo.append(')')
        else:
            if memo[-1]==i:
                memo.pop()
            else:
                return False
    if memo==[]:return True
    else:return False
